### PR TITLE
Use `pd.Period` instead of `pd.Timestamp`.

### PR DESCRIPTION
**Pull request by jaheba**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Please tag this pr with at least one of these labels to make our release process faster:** BREAKING, new feature, bug fix, other change, dev setup

### CODE REVIEW

**Comment by lostella** [comment]
Fixes #1939?

**Comment by jaheba** [comment]
> Fixes #1939?

Yes!

**Comment by kashif** [comment]
@jaheba with this change would it be possible to remove the requirement of specifying `freq` arguments to the forecast generators eg. SampleForecast and QuantileForecasts? I believe the freq can be inferred from the start_date period and this will also help me train models without a freq string specified...

**Comment by jaheba** [comment]
> @jaheba with this change would it be possible to remove the requirement of specifying `freq` arguments to the forecast generators eg. SampleForecast and QuantileForecasts? I believe the freq can be inferred from the start_date period and this will also help me train models without a hard-coded freq string specified...

In short, no, this doesn't change anything about the required `freq` problem.

But I agree, we should make it at least optional, since in many cases `freq` is just carried around because it has to.

**Comment by kashif** [comment]
sorry, I didnt mean we should remove it but yes make it optional since e.g. here: 

https://github.com/awslabs/gluon-ts/blob/master/src/gluonts/model/forecast.py#L507-L508

we should be able to infer the freq from the start period now... thus the freq string can be made optional

**Comment by kashif** [comment]
if you want i can send that PR after you are done with this one?

**Comment by jaheba** [comment]
`Forecast` is something I generally want to rework. I guess we can make it optional first and then think about how to do it right.

I think we should generally remove `start` and use an index in cases where we want to associate a value with a time instead.

File: src/gluonts/data/__init__.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (jaheba) [inline src/gluonts/data/__init__.py:1]: I should remove this file.

File: src/gluonts/core/ty.py [hunk 2]
```diff
@@ -17,6 +17,16 @@
 
 import pydantic
 
+from typing import TypeVar, Optional
+
+T = TypeVar("T")
+
+
+def unwrap(x: Optional[T]) -> T:
```
Comment (jaheba) [inline src/gluonts/core/ty.py:25]: I guess we can add this in a different PR

File: src/gluonts/dataset/common.py [hunk 3]
```diff
@@ -165,10 +150,14 @@ def __init__(
         freq: str,
         one_dim_target: bool = True,
         cache: bool = False,
+        use_timestamp: bool = False,
```
Comment (lostella) [inline src/gluonts/dataset/common.py:153]: This needs to be added to the docstring

File: src/gluonts/dataset/common.py [hunk 4]
```diff
@@ -242,8 +231,10 @@ def __init__(
         data_iter: Iterable[DataEntry],
         freq: str,
         one_dim_target: bool = True,
+        use_timestamp: bool = False,
```
Comment (lostella) [inline src/gluonts/dataset/common.py:234]: Add to docstring

File: src/gluonts/dataset/common.py [hunk 5]
```diff
@@ -288,58 +273,22 @@ class Config:
         arbitrary_types_allowed = True
 
     freq: Union[str, pd.DateOffset]
+    use_timestamp: bool = False
```
Comment (lostella) [inline src/gluonts/dataset/common.py:276]: Add to docstring

File: src/gluonts/dataset/common.py [hunk 6]
```diff
@@ -288,58 +273,22 @@ class Config:
...
     name: str = FieldName.START
-    tz_strategy: TimeZoneStrategy = TimeZoneStrategy.error
 
     def __call__(self, data: DataEntry) -> DataEntry:
         try:
-            timestamp = ProcessStartField.process(data[self.name], self.freq)
+            if self.use_timestamp:
+                data[self.name] = pd.Timestamp(data[self.name])
+            else:
+                data[self.name] = pd.Period(data[self.name], self.freq)
```
Comment (lostella) [inline src/gluonts/dataset/common.py:284]: Why do we need this option?

Comment (jaheba) [reply src/gluonts/dataset/common.py:284]: For point-processes datasets, where `start` needs to be a `Timestamp`. These also have an `end` field which needs to be a timestamp, which we don't handle.

It's a hack.

File: src/gluonts/dataset/artificial/_base.py [hunk 7]
```diff
@@ -143,10 +157,16 @@ def __init__(
         self.is_promotions = is_promotions
         self.holidays = holidays
 
+        if isinstance(self.freq, Week):
+            self.freq = Week(
+                self.freq.n, weekday=pd.Timestamp(start).weekday()
+            )
+        self.start = cast(pd.Period, pd.Period(start, self.freq))
```
[the remark this answers is not in the document]
Comment (jaheba) [reply src/gluonts/dataset/artificial/_base.py:164]: Because pyright complained, because of this:

```py
class Period(PeriodMixin):
    ordinal: int  # int64_t
    freq: BaseOffset

    # error: "__new__" must return a class instance (got "Union[Period, NaTType]")
    def __new__(  # type: ignore[misc]
        cls,
        value=...,
        freq: int | str | BaseOffset | None = ...,
        ordinal: int | None = ...,
        year: int | None = ...,
        month: int | None = ...,
        quarter: int | None = ...,
        day: int | None = ...,
        hour: int | None = ...,
        minute: int | None = ...,
        second: int | None = ...,
    ) -> Period | NaTType: ...
```


https://github.com/pandas-dev/pandas/blob/9289c46e166533dfafd18843137f101624cf41a5/pandas/_libs/tslibs/period.pyi#L62-L79

File: examples/anomaly_detection.py [hunk 8]
```diff
@@ -97,8 +97,8 @@
         i = k // nll.shape[1]
         t = k % nll.shape[1]
 
-        time_index = pd.date_range(
-            pd.Timestamp(data_entry["forecast_start"][i]),
+        time_index = pd.period_range(
+            data_entry["forecast_start"][i],
             periods=context_length + prediction_length,
         )
```
[the remark this answers is not in the document]
Comment (jaheba) [reply examples/anomaly_detection.py:103]: It takes the frequency of the `start` field.

**Review by lostella — APPROVED** [review]

**Comment by kashif** [comment]
@jaheba testing models after this commit I have the feeling that the time feature creation has become a bit slower... meaning its taking longer to make batches for training... the CPU also starts to throttle up when making batches.... did you have that feeling on your end?

**Comment by jaheba** [comment]
Having tested it. I think we now create a new `period_range` for each instance, where we used a cache before. It's something I wanted to address later.

Do you have rough numbers?

**Comment by jaheba** [comment]
Also, I've removed the cache for time-stamp creation which might also slow things down a bit.

**Comment by kashif** [comment]
@jaheba another issue I see is that the ground truth data seems to be a bit screwed up see before traffic:
download-5

and now there seem to be spurious glitches:

download-6

can you kindly confirm you also see this?

**Comment by kashif** [comment]
the data is traffic... i do not believe i am doing anything to the ground truth data with the model... e.g. pytorch deepAR i am just doing the standard prediction and then plotting...

**Comment by jaheba** [comment]
Screenshot 2022-06-08 at 17 08 45

This is how it looks for me.

**Comment by kashif** [comment]
hmmm ok good to know... wonder what's going on at my end.... perhaps i am going crazy....
